flybywiresim/a32nx
# fix: EFB Setting to enable MCDU Server connection to avoid too many connection attempts

## Summary of Changes
EFB Setting to enable MCDU Server connection to avoid too many connection attempts

It has become apparent that the MSFS internal engine does not close and remove websocket objects correctly when the MCDU server is not reachable (e.g. offline). This has led to memory leaks and stutter issues for many users. This issue has been reported to Asobo [6375](https://devsupport.flightsimulator.com/questions/6375/websockets-are-not-cleaned-up.html).

In the meantime this PR offers a workaround for the issue by introducing an EFB Setting to enable and disable the MCDU Server connection.

The EFB Setting has three options: "Auto On", "Auto Off", "Perm Off".

Auto On: Will try to connect to the MCDU Server with each start of a flight. The number of attempts to connect to a MCDU server is limited to 60 (=5min) after which the EFB setting will be disabled ("Auto Off"). 
The user can enable the setting immediately again to get another 5min of connection attempts. 

Auto Off: After 60 attempts the setting will jump to this setting. If this is the seeting from the last flight it will automatically set to "Auto On" to attempt to connect. 

Perm Off: The MCDU will never attempt to connect to the MCDU Server. 

Note: Although during the 5min of connection attempts there will be some accumulation of websocket objects this is a good compromise between mitigating the issue and comfort for the user. Especially as users reported issues only after a longer period of flight. 

Linked Docs PR: docs update: https://github.com/flybywiresim/docs/pull/499

## Screenshots
![image](https://user-images.githubusercontent.com/16833201/159256787-eabecb6e-2f2d-4b14-bf93-57014159da50.png)

The error messages are expected - although not nice they can't be avoided when trying to make sure all sockets are closed. 
![image](https://user-images.githubusercontent.com/16833201/159382428-fa0ccc63-7dc5-4ea7-aa47-bb95c5ff39ea.png)

![image](https://user-images.githubusercontent.com/16833201/159256906-50e05e13-8efc-4f84-bd1f-6e742e9772d8.png)

Discord username (if different from GitHub): Cr_Maverick#6475

## Testing instructions

- Check the new EFB Setting  "Enable MCDU Server Connection" and enable it ("Auto On") if not already the case.
- With the MCDU Server OFF wait for 5min and confirm that the new setting will be deactivated ("Auto Off"). 
	- (you can monitor the progress in the Coherent GT Debugger if you know how to)
- Try to start the MCDU Server - it should not be able to connect to the aircraft
- Enable the EFB setting again ("Auto On") - the MCDU Server should no connect to the aircraft
- Check that the remote Web MCDU Interface works as before
- Disable the EFB setting - the MCDU Server should loose connection
- Test "Perm Off" setting - restart the flight (do a new flight to be sure or even restart the sim)
- Test any other combination you might think of
- Monitor the FPS counter and general smoothness of the sim is as before

Find the MCDU Server docs here: 
https://docs.flybywiresim.com/fbw-a32nx/feature-guides/web-mcdu/


## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page
